doc/developer/design/20250321_password_authentication.md

```diff
@@ -0,0 +1,128 @@
+
+**Password Rotation or Expiration**: While not currently in scope, the design should allow for additions in the future, such as password rotation and expiration.
+
+**Abuse Detection**: Action should be taken action when brute force attempts are detected, either by locking the user or blocking the requesting client IP.
+
+### 2. Password Management Syntax:
+
+Users should be able to manage their passwords and set passwords for roles they create. Superusers should be able to alter any non-internal user's password.
+
+-   **Syntax Examples**:
+
+    -   `CREATE ROLE hunter WITH PASSWORD 'hunter2';`
+    -   `ALTER ROLE hunter SET (PASSWORD 'hunter2');`
+    -   `ALTER ROLE user_name PASSWORD NULL`
+
+-   **Password Policies**: We will not implement password policies but should enforce a configurable minimum password length.
+
+-   **Login Role Determination**: We should display whether a role has a password and can be used as a login role in `mz_roles`. See [pg_roles](https://www.postgresql.org/docs/9.1/view-pg-roles.html).
+
+### 3. Hashed Password Storage
+
+-   **Secure Hashing**: For future compatibility with SCRAM, passwords should be hashed using `scram-sha-256` encryption. The RFC for [SCRAM](https://datatracker.ietf.org/doc/html/rfc5802) dictates an iteration count of at least 4096. However, the RFC for SCRAM-SHA-256 dictates that "[the hash iteration-count should be such that a modern machine will take 0.1 seconds to perform the complete algorithm](https://www.rfc-editor.org/rfc/rfc7677)". This also dictates at least 4096 with a recommendation of 15000. We should aim for 15000 if it does not negatively impact connection times or load.
+
+-   **Access Control**: Hashed passwords should not be extracted from the secret store and should not be accessible through queryable tables. Passwords must not be logged or output in stack traces.
+
+-   **Password Dating**: The creation date of passwords will be recorded along with the hashed and salted password. When the password changes a creation date should be reset. The creation date of passwords should be queryable.
+
+-   **OPTIONAL: Password Versioning**: Updates to the password hashing mechanisms may be required. As long as we are receiving passwords in plain text we should be able to take a validated password and replace the existing a new securely hashed value. This may require prefixing the password with data about the hash algorithm or parameters.
+
+### 4. Configurable admin system login:
+
+Passwords for `mz_system` and `mz_support` roles will be settable via environment variables `MZ_MZ_<USER>_EXTERNAL_LOGIN_PASSWORD`. Orchestratord should provide a set of parameters to set these variables via a Kubernetes secret. Additionally, We should enable login of system users through the external ports when they have external login passwords set. Login through the external port must not be possible unless this flag is set, this logic should not rely on whether the internal user has a password.
```

Comment:
Nit: Should be `MZ_<USER>_EXTERNAL_LOGIN_PASSWORD` given `mz_system` would give `MZ_MZ_MZ_SYSTEM_EXTERNAL_LOGIN_PASSWORD`

Reply:
Maybe the user should be last? `MZ_EXTERNAL_LOGIN_PASSWORD_<USER>`

Reply:
Sounds good!

`EXTERNAL_LOGIN_PASSWORD_<USER>`

with a env variable of 

`MZ_EXTERNAL_LOGIN_PASSWORD_<USER>`